File: 2024/01/01.go
package main

import (
	"bufio"
	"fmt"
	"math"
	"os"
	"sort"
	"strconv"
	"strings"
)

func main() {

	// Read the input file
	file, err := os.Open("input.txt")
	if err != nil {
		fmt.Println("Error reading file")
		return
	}
	defer file.Close()

	// Parse the input file
	lines := []string{}
	col_1 := []int{}
	col_2 := []int{}
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		lines = append(lines, scanner.Text())
		line := strings.Split(scanner.Text(), "   ")
		num1, _ := strconv.Atoi(line[0])
		num2, _ := strconv.Atoi(line[1])
		col_1 = append(col_1, num1)
		col_2 = append(col_2, num2)
	}

	// Print the results from each part
	fmt.Println(part1(col_1, col_2))
	fmt.Println(part2(col_1, col_2))

	// Check for errors
	if err := scanner.Err(); err != nil {
		fmt.Println("Error reading file")
		return
	}
}

func part1(col_1 []int, col_2 []int) int {

	sort.Ints(col_1)
	sort.Ints(col_2)

	distances := []int{}

	for i := 0; i < len(col_1); i++ {
		distances = append(distances, int(math.Abs(float64(col_1[i]-col_2[i]))))
	}

	return sum(distances)
}

func part2(col_1 []int, col_2 []int) int {

	similarities := []int{}

	for i := 0; i < len(col_1); i++ {
		total := 0
		curr_num := col_1[i]
		for j := 0; j < len(col_2); j++ {
			if curr_num == col_2[j] {
				total += 1
			}
		}
		similarities = append(similarities, (total * curr_num))
	}

	return sum(similarities)
}

func sum(arr []int) int {
	total := 0
	for _, val := range arr {
		total += val
	}
	return total
}

File: 2024/01/01.py
# Read the input file
input_file = open("input.txt", "r")
lines = input_file.readlines()

lines = [line.strip().split() for line in lines]
lines = [[int(x) for x in pair] for pair in lines]
col_1 = [pair[0] for pair in lines]
col_2 = [pair[1] for pair in lines]


# Part 1
# Each line is set up like "123 456\n"
# Need to make a list of each column from the pairs of each line
def part1():
  col_1.sort()
  col_2.sort()

  distances = []

  for i in range(len(col_1)):
      x = col_1[i]
      y = col_2[i]
      distances.append(abs(x - y))

  return sum(distances)


# Part 2
# We will need to find how many times the number form col_1 appears in col_2
def part2():
  # iterate through col 1
  # find how many times num is in col 2

  similarity = []

  for num in col_1:
    total = col_2.count(num)
    similarity.append(total * num)

  return sum(similarity)



if __name__ == "__main__":
    print(part1())
    print(part2())